File: Microsoft.Toolkit.Uwp.Input.GazeInteraction/GazeFeedbackPopupFactory.cpp
//Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license.
//See LICENSE in the project root for license information.

#include "pch.h"
#include "GazeFeedbackPopupFactory.h"
#include "GazeInput.h"

BEGIN_NAMESPACE_GAZE_INPUT

Popup GazeFeedbackPopupFactory::Get()
{
    Popup popup;
    winrt::Microsoft::UI::Xaml::Shapes::Rectangle rectangle;

    if (s_cache.size != 0)
    {
        popup = s_cache[0];
        s_cache.erase(s_cache.begin());

        rectangle = popup.Child;
    }
    else
    {
        popup = Popup();

        rectangle = winrt::Microsoft::UI::Xaml::Shapes::Rectangle();
        rectangle.IsHitTestVisible = false;

        popup.Child = rectangle;
    }

    rectangle.StrokeThickness = GazeInput::DwellStrokeThickness;

    return popup;
}

void GazeFeedbackPopupFactory::Return(Popup popup)
{
    popup.IsOpen = false;
    s_cache.push_back(popup);
}

END_NAMESPACE_GAZE_INPUT

File: CommunityToolkit.WinUI.Input.GazeInteraction/GazeFeedbackPopupFactory.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Collections.Generic;
using Microsoft.UI.Xaml.Controls.Primitives;

namespace CommunityToolkit.WinUI.Input.GazeInteraction
{
    internal class GazeFeedbackPopupFactory
    {
        private readonly List<Popup> _cache = new List<Popup>();

        public Popup Get()
        {
            Popup popup;
            Microsoft.UI.Xaml.Shapes.Rectangle rectangle;

            if (_cache.Count != 0)
            {
                popup = _cache[0];
                _cache.RemoveAt(0);

                rectangle = popup.Child as Microsoft.UI.Xaml.Shapes.Rectangle;
            }
            else
            {
                popup = new Popup();

                rectangle = new Microsoft.UI.Xaml.Shapes.Rectangle
                {
                    IsHitTestVisible = false
                };

                popup.Child = rectangle;
            }

            rectangle.StrokeThickness = GazeInput.DwellStrokeThickness;

            return popup;
        }

        public void Return(Popup popup)
        {
            popup.IsOpen = false;
            _cache.Add(popup);
        }
    }
}